rpc/storage.go

```diff
@@ -0,0 +1,273 @@
+
+/****************************************************
+		Storage Handlers
+*****************************************************/
+
+// StorageAt gets the value of the storage at the given address and key.
+//
+// It follows the specification defined here:
+// https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_api_openrpc.json#L110
+func (h *Handler) StorageAt(address, key felt.Felt, id BlockID) (*felt.Felt, *jsonrpc.Error) {
+	stateReader, stateCloser, rpcErr := h.stateByBlockID(&id)
+	if rpcErr != nil {
+		return nil, rpcErr
+	}
+	defer h.callAndLogErr(stateCloser, "Error closing state reader in getStorageAt")
+
+	value, err := stateReader.ContractStorage(&address, &key)
+	if err != nil {
+		return nil, ErrContractNotFound
+	}
+
+	return value, nil
+}
+
+// StorageProof returns the merkle paths in one of the state tries: global state, classes, individual contract
+//
+// It follows the specification defined here:
+// https://github.com/starkware-libs/starknet-specs/blob/647caa00c0223e1daab1b2f3acc4e613ba2138aa/api/starknet_api_openrpc.json#L910
+func (h *Handler) StorageProof(id BlockID, classes, contracts []felt.Felt, storageKeys []StorageKeys) (*StorageProofResult, *jsonrpc.Error) {
+	stateReader, stateCloser, err := h.bcReader.HeadState()
+	if err != nil {
+		return nil, ErrInternal.CloneWithData(err)
+	}
+	defer h.callAndLogErr(stateCloser, "Error closing state reader in getStorageProof")
+
+	head, err := h.bcReader.Head()
+	if err != nil {
+		return nil, ErrInternal.CloneWithData(err)
+	}
+
+	if !(id.Latest || head.Number == id.Number) {
+		return nil, ErrBlockNotRecentForProof
+	}
+
+	storageRoot, classRoot, err := stateReader.StateAndClassRoot()
+	if err != nil {
+		return nil, ErrInternal.CloneWithData(err)
+	}
+
+	result := &StorageProofResult{
+		GlobalRoots: &GlobalRoots{
+			ContractsTreeRoot: storageRoot,
+			ClassesTreeRoot:   classRoot,
+			BlockHash:         head.Hash,
+		},
+	}
+
+	result.ClassesProof, err = getClassesProof(stateReader, classes)
+	if err != nil {
+		return nil, ErrInternal.CloneWithData(err)
+	}
+	result.ContractsProof, err = getContractsProof(stateReader, contracts)
+	if err != nil {
+		return nil, ErrInternal.CloneWithData(err)
+	}
+	result.ContractsStorageProofs, err = getContractsStorageProofs(stateReader, storageKeys)
+	if err != nil {
+		return nil, ErrInternal.CloneWithData(err)
+	}
+
+	return result, nil
+}
+
+// StorageKeys represents an item in `contracts_storage_keys. parameter
+// https://github.com/starkware-libs/starknet-specs/blob/647caa00c0223e1daab1b2f3acc4e613ba2138aa/api/starknet_api_openrpc.json#L938
+type StorageKeys struct {
+	Contract felt.Felt   `json:"contract_address"`
+	Keys     []felt.Felt `json:"storage_keys"`
+}
+
+// MerkleNode represents a proof node in a trie
+// https://github.com/starkware-libs/starknet-specs/blob/647caa00c0223e1daab1b2f3acc4e613ba2138aa/api/starknet_api_openrpc.json#L3632
+type MerkleNode interface{}
```

Comment:
Maybe we can constrain this somehow, eg something like GetNodeType() NodeType??

Reply:
Or merge into a single type??

Reply:
Yes, I was thinking about some artificial function - but now I have and idea: conversion to our ProofNode